core/thread/key.go

```diff
@@ -0,0 +1,122 @@
+package thread
+
+import (
+	"fmt"
+
+	mbase "github.com/multiformats/go-multibase"
+	sym "github.com/textileio/go-threads/crypto/symmetric"
+)
+
+var (
+	// Indicates an invalid byte slice was given to KeyFromBytes.
+	ErrInvalidKey = fmt.Errorf("invalid key")
+)
+
+// Key is a thread encryption key with two components.
+// Service key is used to encrypt outer log record linkages.
+// Read key is used to encrypt inner record events.
+type Key struct {
+	sk *sym.Key
+	rk *sym.Key
+}
+
+// NewKey wraps service and read keys.
+func NewKey(sk, rk *sym.Key) Key {
+	if sk == nil {
+		panic("service-key must not be nil")
+	}
+	return Key{sk: sk, rk: rk}
+}
+
+// NewServiceKey wraps a service-only key.
+func NewServiceKey(sk *sym.Key) Key {
+	return Key{sk: sk}
+}
+
+// NewRandomKey returns a random key, which includes a service and read key.
+func NewRandomKey() Key {
+	return Key{sk: sym.New(), rk: sym.New()}
+}
+
+// NewRandomServiceKey returns a random service-only key.
+func NewRandomServiceKey() Key {
+	return Key{sk: sym.New()}
+}
+
+// KeyFromBytes returns a key by wrapping k.
+func KeyFromBytes(b []byte) (k Key, err error) {
+	if len(b) != sym.KeyBytes && len(b) != sym.KeyBytes*2 {
+		return k, ErrInvalidKey
+	}
+	sk, err := sym.FromBytes(b[:sym.KeyBytes])
+	if err != nil {
+		return k, err
+	}
+	var rk *sym.Key
+	if len(b) == sym.KeyBytes*2 {
+		rk, err = sym.FromBytes(b[sym.KeyBytes:])
+		if err != nil {
+			return k, err
+		}
+	}
+	return Key{sk: sk, rk: rk}, nil
+}
+
+// KeyFromString returns a key by decoding a base58-encoded string.
+func KeyFromString(s string) (k Key, err error) {
+	_, b, err := mbase.Decode(s)
+	if err != nil {
+		return k, err
+	}
+	return KeyFromBytes(b)
+}
+
+// Service returns the service key.
+func (k Key) Service() *sym.Key {
+	return k.sk
+}
+
+// Read returns the read key.
+func (k Key) Read() *sym.Key {
```

Comment:
Totally agree that `Key` suffix is redundant, but looks weird. :)
I still think `Read` is the correct name though.

Reply:
Hmm, right. Sounds like _read_ a `io.Reader`.

Reply:
Yeah, but just biased from the Go ecosystem. The coincidence of names!